pyat/at/physics/rdt.py

```diff
@@ -0,0 +1,545 @@
+
+    if (RDTType.FOCUSING in rdttype) or (RDTType.ALL in rdttype):
+        b2lm = b2l[mask_b2l]
+        betaxm = betax[mask_b2l]
+        betaym = betay[mask_b2l]
+        pxm = px[mask_b2l]
+        pym = py[mask_b2l]
+        rdts["h20000"] = sum((b2lm / 8) * betaxm * pxm * pxm * pf(2, 0))
+        rdts["h00200"] = sum((b2lm / 8) * betaym * pym * pym * pf(0, 2))
+
+    if (RDTType.COUPLING in rdttype) or (RDTType.ALL in rdttype):
+        a2lm = a2l[mask_a2l]
+        rbetaxm = rbetax[mask_a2l]
+        rbetaym = rbetay[mask_a2l]
+        pxm = px[mask_a2l]
+        pym = py[mask_a2l]
+        rdts["h10010"] = sum((a2lm / 4) * rbetaxm * rbetaym * pxm / pxm * pf(1, -1))
+        rdts["h10100"] = sum((a2lm / 4) * rbetaxm * rbetaym * pxm / pym * pf(1, 1))
+
+    if (RDTType.CHROMATIC in rdttype) or (RDTType.ALL in rdttype):
+        mask_b23l = mask_b2l | mask_b3l
+        b2lm = b2l[mask_b23l]
+        b3lm = b3l[mask_b23l]
+        betaxm = betax[mask_b23l]
+        rbetaxm = rbetax[mask_b23l]
+        betaym = betay[mask_b23l]
+        etaxm = etax[mask_b23l]
+        pxm = px[mask_b23l]
+        pym = py[mask_b23l]
+        rdts["h11001"] = sum((b3lm * betaxm * etaxm / 2 - b2lm * betaxm / 4) * nperiods)
+        rdts["h00111"] = sum((b2lm * betaym / 4 - b3lm * betaym * etaxm / 2) * nperiods)
+        rdts["h20001"] = sum(
+            (b3lm * betaxm * etaxm / 2 - b2lm * betaxm / 4) / 2 * pxm * pxm * pf(2, 0)
+        )
+        rdts["h00201"] = sum(
+            (b2lm * betaym / 4 - b3lm * betaym * etaxm / 2) / 2 * pym * pym * pf(0, 2)
+        )
+        # fmt: off
+        rdts["h10002"] = sum(
+            (b3lm * rbetaxm * etaxm * etaxm - b2lm * rbetaxm * etaxm)
+            / 2 * pxm * pf(1, 0)
+        )
+        # fmt: on
+
+    if (RDTType.GEOMETRIC1 in rdttype) or (RDTType.ALL in rdttype):
+        b3lm = b3l[mask_b3l]
+        betaxm = betax[mask_b3l]
+        betaym = betay[mask_b3l]
+        rbetaxm = rbetax[mask_b3l]
+        pxm = px[mask_b3l]
+        pym = py[mask_b3l]
+        rdts["h21000"] = sum(b3lm * rbetaxm * betaxm / 8 * pxm * pf(1, 0))
+        rdts["h30000"] = sum(b3lm * rbetaxm * betaxm / 24 * pxm * pxm * pxm * pf(3, 0))
+        rdts["h10110"] = sum(-b3lm * rbetaxm * betaym / 4 * pxm * pf(1, 0))
+        rdts["h10020"] = sum(
+            -b3lm * rbetaxm * betaym / 8 * pxm * numpy.conj(pym * pym) * pf(1, -2)
+        )
+        rdts["h10200"] = sum(-b3lm * rbetaxm * betaym / 8 * pxm * pym * pym * pf(1, 2))
+
+    if (RDTType.TUNESHIFT in rdttype) or (RDTType.ALL in rdttype):
+        b4lm = b4l[mask_b4l]
+        betaxm = betax[mask_b4l]
+        betaym = betay[mask_b4l]
+        rdts["dnux_dJx"] = sum(3 * b4lm * betaxm * betaxm / (8 * numpy.pi) * nperiods)
+        rdts["dnux_dJy"] = sum(-3 * b4lm * betaxm * betaym / (4 * numpy.pi) * nperiods)
+        rdts["dnuy_dJy"] = sum(3 * b4lm * betaym * betaym / (8 * numpy.pi) * nperiods)
+
+    if (RDTType.GEOMETRIC2 in rdttype) or (RDTType.ALL in rdttype):
```

Comment:
The geometric rdts of octupoles agree with my theoretical calculation.
There is yet no reference on where are they taken from.
Also, there is not explanation on the difference between GEOMETRIC1 and GEOMETRIC2 RDTTYPEs. Improve the DocString.

I attach the script I wrote to calculate the geometric RDTs of octupoles in Maxima.
If anybody would like also to check it, you could run it as 
```bash
maxima -q -b fourth_order.mc
```
[fourth_order.mc.txt](https://github.com/user-attachments/files/16910908/fourth_order.mc.txt)
Maxima https://maxima.sourceforge.io/docs/manual/maxima.html